parquet-key-management/src/crypto_factory.rs

```diff
@@ -735,6 +736,52 @@ mod tests {
         }
     }
 
+    /// Test caching of key encryption keys when encrypting files
+    #[test]
+    fn test_encryption_key_encryption_key_caching() {
+        let time_controller = crate::kms_manager::mock_time::time_controller();
+
+        let kms_config = Arc::new(KmsConnectionConfig::default());
+        let encryption_config = EncryptionConfigurationBuilder::new("kf".to_owned())
+            .set_double_wrapping(true)
+            .add_column_key("kc1".to_owned(), vec!["x0".to_owned(), "x1".to_owned()])
+            .add_column_key("kc2".to_owned(), vec!["x2".to_owned(), "x3".to_owned()])
+            .set_cache_lifetime(Some(Duration::from_secs(600)))
+            .build();
+
+        let kms_factory = Arc::new(TestKmsClientFactory::with_default_keys());
+        let crypto_factory = CryptoFactory::new(kms_factory.clone());
+
+        let generate_encryption_props = || {
+            let _ = crypto_factory
+                .file_encryption_properties(kms_config.clone(), &encryption_config)
+                .unwrap();
+        };
+
+        assert_eq!(0, kms_factory.keys_wrapped());
+
+        generate_encryption_props();
+        // 1 key wrapped per master key id
+        assert_eq!(3, kms_factory.keys_wrapped());
+
+        time_controller.advance(Duration::from_secs(599));
+        generate_encryption_props();
+        assert_eq!(3, kms_factory.keys_wrapped());
+
+        time_controller.advance(Duration::from_secs(1));
+        generate_encryption_props();
+        assert_eq!(6, kms_factory.keys_wrapped());
+
+        // Refreshing the access token should invalidate the KEK write cache
+        kms_config.refresh_key_access_token("new_secret".to_owned());
+        generate_encryption_props();
+        assert_eq!(9, kms_factory.keys_wrapped());
```

Comment:
I guess this makes sense, but shouldn't the cache discard old/unused keys? I haven't looked at the source fully here.

Reply:
OK. I guess keys are only expired on a read call?  I think it would make sense to enhance this test to show keys expiring.

Reply:
I think the confusing thing here might be what `kms_factory.keys_wrapped()` represents. This isn't the size of the cache but is the number of times the KMS had to be used to wrap a key encryption key. So it is monotonic.

I've added some comments to hopefully clarify what's going on and have also added checks of the cache size.